server/src/main/java/io/deephaven/server/barrage/BarrageMessageProducer.java

```diff
@@ -1140,106 +1149,115 @@ private void updateSubscriptionsSnapshotAndPropagate() {
+                    BitSet addedCols = (BitSet) subscription.targetColumns.clone();
+                    addedCols.andNot(subscription.subscribedColumns);
 
-                        final boolean viewportValid =
-                                subscription.reverseViewport == subscription.targetReverseViewport
-                                        && missingColumns.isEmpty();
+                    final boolean viewportValid = subscription.reverseViewport == subscription.targetReverseViewport
+                            && addedCols.isEmpty();
 
-                        // if we have a valid viewport, can exclude current viewport from requested rows
-                        if (viewportValid && subscription.viewport != null) {
-                            missingRows.remove(subscription.viewport);
-                        }
+                    // don't request valid rows that the client already possesses
+                    if (viewportValid && subscription.viewport != null) {
+                        subscription.growingRemainingViewport.remove(subscription.viewport);
+                    }
 
-                        // remove rows that we already have selected for the upcoming snapshot
-                        missingRows.remove(subscription.targetReverseViewport ? reverseSnapshotRowSet : snapshotRowSet);
+                    // get the current set for this viewport direction
+                    final WritableRowSet currentSet =
+                            subscription.targetReverseViewport ? reverseSnapshotRowSet : snapshotRowSet;
 
-                        // shrink this to the `rowsRemaining` value
-                        if (missingRows.size() > rowsRemaining) {
-                            final long key = missingRows.get(rowsRemaining);
-                            missingRows.removeRange(key, Long.MAX_VALUE - 1);
-                        }
+                    // get the rows that we need that are already in the snapshot
+                    try (final RowSet overlap = subscription.growingRemainingViewport.extract(currentSet)) {
+                        if (rowsRemaining <= 0) {
+                            // we can't request additional rows, but can use other rows in this snapshot
+                            subscription.growingIncrementalViewport = overlap.copy();
+                        } else {
+                            // request as many rows as we are allowed
+                            try (final WritableRowSet additional = subscription.growingRemainingViewport.copy()) {
 
-                        subscription.snapshotViewport = viewportValid && subscription.viewport != null
-                                ? subscription.viewport.copy()
-                                : RowSetFactory.empty();
+                                // shrink the set of new rows to <= `rowsRemaining` size
+                                if (additional.size() > rowsRemaining) {
```

Comment:
Peform this check against `subscription.growingIncrementalViewport` and you can eliminate a copy when `rowsRemaining` is sufficiently large to fulfill the entire viewport

Reply:
I think this is not worth the effort to avoid a single fairly efficient copy.  Much of the code would need duplicated along each path and maintenance and readability are concerns already.